lib/processPbrMetallicRoughness.js

```diff
@@ -0,0 +1,720 @@
+        if (defined(parameterValues.metallicFactor)) {
+            fragmentShader += '    metalness *= u_metallicFactor;\n';
+        }
+        if (defined(parameterValues.roughnessFactor)) {
+            fragmentShader += '    roughness *= u_roughnessFactor;\n';
+        }
+    } else {
+        if (defined(parameterValues.metallicFactor)) {
+            fragmentShader += '    float metalness = clamp(u_metallicFactor, 0.0, 1.0);\n';
+        } else {
+            fragmentShader += '    float metalness = 1.0;\n';
+        }
+        if (defined(parameterValues.roughnessFactor)) {
+            fragmentShader += '    float roughness = clamp(u_roughnessFactor, 0.04, 1.0);\n';
+        } else {
+            fragmentShader += '    float roughness = 1.0;\n';
+        }
+    }
+    fragmentShader += '    vec3 v = -normalize(v_positionEC);\n';
+
+    // Generate fragment shader's lighting block
+    fragmentShader += '    vec3 lightColor = vec3(1.0, 1.0, 1.0);\n';
+    if (optimizeForCesium) {
+        fragmentShader += '    vec3 l = normalize(czm_sunDirectionEC);\n';
+    } else {
+        fragmentShader += '    vec3 l = vec3(0.0, 0.0, 1.0);\n';
+    }
+    fragmentShader += '    vec3 h = normalize(v + l);\n';
+    fragmentShader += '    vec3 r = normalize(reflect(v, n));\n';
+    fragmentShader += '    float NdotL = clamp(dot(n, l), 0.001, 1.0);\n';
+    fragmentShader += '    float NdotV = abs(dot(n, v)) + 0.001;\n';
+    fragmentShader += '    float NdotH = clamp(dot(n, h), 0.0, 1.0);\n';
+    fragmentShader += '    float LdotH = clamp(dot(l, h), 0.0, 1.0);\n';
+    fragmentShader += '    float VdotH = clamp(dot(v, h), 0.0, 1.0);\n';
+
+    fragmentShader += '    vec3 f0 = vec3(0.04);\n';
+    fragmentShader += '    float alpha = roughness * roughness;\n';
+    fragmentShader += '    vec3 diffuseColor = baseColor * (1.0 - metalness);\n';
+    fragmentShader += '    vec3 specularColor = mix(f0, baseColor, metalness);\n';
+    fragmentShader += '    float reflectance = max(max(specularColor.r, specularColor.g), specularColor.b);\n';
+    fragmentShader += '    vec3 r90 = vec3(clamp(reflectance * 25.0, 0.0, 1.0));\n';
+    fragmentShader += '    vec3 r0 = specularColor.rgb;\n';
+
+    fragmentShader += '    vec3 F = fresnelSchlick2(r0, r90, VdotH);\n';
+    fragmentShader += '    float G = smithVisibilityGGX(alpha, NdotL, NdotV);\n';
+    fragmentShader += '    float D = GGX(alpha, NdotH);\n';
+
+    fragmentShader += '    vec3 diffuseContribution = (1.0 - F) * lambertianDiffuse(baseColor);\n';
+    fragmentShader += '    vec3 specularContribution = F * G * D / (4.0 * NdotL * NdotV);\n';
+    fragmentShader += '    vec3 color = NdotL * lightColor * (diffuseContribution + specularContribution);\n';
+
+    if (optimizeForCesium) {
+        fragmentShader += '    vec3 diffuseIrradiance = vec3(0.5);\n';
+        fragmentShader += '    vec3 specularIrradiance = textureCube(czm_cubeMap, r).rgb;\n';
```

Comment:
I should probably clarify the name, but I added the specular environment map as a Cesium builtin for now since anyone using Cesium will use the same map.